### ISSUE TITLE
Show TouchBar on Fullscreen mode?

### ISSUE THREAD

**Issue by hankisthatyou** [bikkelbroeders/TouchBarDemoApp#9]
Not sure if this is achievable, but would be nice if we can have the TouchBar shown on fullscreen mode.

**Comment by robbertkl** [comment]
Do you mean for the on-screen Touch Bar, or the USB iPad Touch Bar? What do you mean by fullscreen mode?

Currently, the Fn-toggled on-screen Touch Bar already works in fullscreen macOS apps, but perhaps this is not what you mean.

**Comment by hankisthatyou** [comment]
When switching between windows and full screen apps, currently you need to retoggle on-screen Touch Bar. Is there anyway to make it stay in the same position all the time?

**Comment by avaidyam** [comment]
There is; I've got a very stripped down version of this tool that does that. I'll make the code happy and release it later. It also supports docking the window at the bottom of the screen instead of following your cursor around.

**Comment by robbertkl** [comment]
Great, we're also working on something similar, but I'm interested to see what you've made @avaidyam.

**Comment by avaidyam** [comment]
In `[TouchBarWindow init...]`:

```
        // For rounded edges:
        self.styleMask = NSTitledWindowMask | NSFullSizeContentViewWindowMask;
        self.titlebarAppearsTransparent = YES;
        self.titleVisibility = NSWindowTitleHidden;
        [self standardWindowButton:NSWindowCloseButton].hidden = YES;
        [self standardWindowButton:NSWindowFullScreenButton].hidden = YES;
        [self standardWindowButton:NSWindowZoomButton].hidden = YES;
        [self standardWindowButton:NSWindowMiniaturizeButton].hidden = YES;

        // For squared edges:
        //self.styleMask = NSBorderlessWindowMask;

        // Common:
        self.movable = NO;
        self.acceptsMouseMovedEvents = YES;
        self.movableByWindowBackground = NO;
        self.level = CGWindowLevelForKey(kCGAssistiveTechHighWindowLevelKey);
        self.collectionBehavior = (NSWindowCollectionBehaviorCanJoinAllSpaces | NSWindowCollectionBehaviorStationary | NSWindowCollectionBehaviorIgnoresCycle | NSWindowCollectionBehaviorFullScreenDisallowsTiling);
        self.backgroundColor = [NSColor blackColor];
        [self _setPreventsActivation:YES];
        [self setFrame:NSMakeRect(0, 0, 1085 + 10, 30 + 10) display:YES];
        self.contentView = [TouchBarView new];
```

**Comment by avaidyam** [comment]
In `AppDelegate`:

```
// FIXME: Currently has a shadow gap between the bottom edges.
- (NSPoint)screenCenterOrigin {
    NSPoint mousePoint = [NSEvent mouseLocation];
    NSScreen* currentScreen = [NSScreen mainScreen];
    for(NSScreen* screen in [NSScreen screens]) {
        if(NSPointInRect(mousePoint, screen.visibleFrame)) {
            currentScreen = screen;
            break;
        }
    }

    NSRect screenFrame = currentScreen.visibleFrame;
    CGFloat screenDiff = screenFrame.size.width - _touchBarWindow.frame.size.width;
    return NSMakePoint(screenDiff / 2.0, 0.0);
}

- (NSPoint)mouseTouchOrigin {
    NSPoint mousePoint = [NSEvent mouseLocation];
    NSScreen* currentScreen = [NSScreen mainScreen];
    for(NSScreen* screen in [NSScreen screens]) {
        if(NSPointInRect(mousePoint, screen.visibleFrame)) {
            currentScreen = screen;
            break;
        }
    }

    NSRect screenFrame = currentScreen.visibleFrame;
    CGFloat screenRight = screenFrame.origin.x + screenFrame.size.width;
    CGFloat barWidth = _touchBarWindow.frame.size.width;
    CGFloat barHeight = _touchBarWindow.frame.size.height;

    NSPoint origin = mousePoint;
    origin.x -= barWidth * 0.5;
    origin.y -= barHeight;

    if(origin.x < screenFrame.origin.x)
        origin.x = screenFrame.origin.x;

    if(origin.x + barWidth > screenRight)
        origin.x = screenRight - barWidth;

    if(origin.y - barHeight < screenFrame.origin.y)
        origin.y += barHeight;
    return origin;
}
```

Then, where you were doing the `setFrameOrigin:` before,

```
                    NSUserDefaultsController *d2 = [NSUserDefaultsController sharedUserDefaultsController];
                    BOOL origin = [[d2 valueForKeyPath:@"values.FollowMouse"] boolValue];
                    [_touchBarWindow setFrameOrigin:origin ? self.mouseTouchOrigin : self.screenCenterOrigin];
```

**Comment by avaidyam** [comment]
@robbertkl @AndreasVerhoeven That covers the majority of the changes I'd made. I ended up gutting out the iOS portion on a whim. The changes above allow the bar to both follow the mouse cursor and stay docked to the bottom of the screen, and stay on top of every window (except the login window).

**Comment by avaidyam** [comment]
I think this issue is fixed with the recent commit 😉

**Comment by robbertkl** [comment]
Thanks @avaidyam for the changes. I took the rounded edges and other window fixes (level, collectionBehavior, etc). It will be included in the next release.

I did not yet take the "stick to bottom" code, as we're planning something a bit different (though similar).

Next time, just send us a PR (for each change/feature separately), so we can apply it more easily and your name/commit gets into the history.
